File: Med Spa App/ml-models/src/features.py
"""Feature extraction from Supabase data. Maps to TS feature extractors in packages/intelligence."""

from datetime import datetime, timedelta
from typing import Any


def extract_appointment_features(appointments: list[dict]) -> list[float]:
    """Extract appointment features. Mirrors appointmentFeatures in TS."""
    total = len(appointments)
    completed = sum(1 for a in appointments if a.get("status") == "completed")
    cancelled = sum(1 for a in appointments if a.get("status") == "cancelled")
    no_shows = sum(1 for a in appointments if a.get("status") == "cancelled" and not a.get("intake_completed"))
    completion_rate = completed / total if total > 0 else 0
    cancel_rate = cancelled / total if total > 0 else 0
    no_show_rate = no_shows / total if total > 0 else 0
    return [total, completed, cancelled, no_shows, completion_rate, cancel_rate, no_show_rate]


def extract_payment_features(payments: list[dict]) -> list[float]:
    """Extract payment features. Mirrors paymentFeatures in TS."""
    total = len(payments)
    completed = sum(1 for p in payments if p.get("status") == "completed")
    failed = sum(1 for p in payments if p.get("status") == "failed")
    total_revenue = sum(p.get("amount", 0) for p in payments if p.get("status") == "completed")
    avg_payment = total_revenue / completed if completed > 0 else 0
    return [total, completed, failed, round(total_revenue, 2), round(avg_payment, 2)]


def extract_engagement_features(audit_logs: list[dict]) -> list[float]:
    """Extract engagement features. Mirrors engagementFeatures in TS."""
    login_count = sum(1 for l in audit_logs if l.get("action") == "login")
    login_logs = [l for l in audit_logs if l.get("action") == "login"]
    if login_logs:
        latest = max(l.get("created_at", "") for l in login_logs)
        if latest:
            days_since_login = (datetime.now() - datetime.fromisoformat(latest.replace("Z", "+00:00"))).days
        else:
            days_since_login = -1
    else:
        days_since_login = -1
    return [login_count, days_since_login, len(audit_logs)]


def build_feature_matrix(appointments: list[dict], payments: list[dict], audit_logs: list[dict]) -> list[float]:
    """Build complete feature vector from all extractors."""
    features = []
    features.extend(extract_appointment_features(appointments))
    features.extend(extract_payment_features(payments))
    features.extend(extract_engagement_features(audit_logs))
    return features

File: Med Spa App/packages/intelligence/src/predictions/features.ts
import type { FeatureExtractor } from './types';
import type { RuleContext } from '../types';

export const appointmentFeatures: FeatureExtractor = {
  name: 'appointments',
  extract(ctx: RuleContext): number[] {
    const patientAppts = ctx.customerId
      ? ctx.appointments.filter((a) => a.patient_id === ctx.customerId)
      : ctx.appointments;

    const total = patientAppts.length;
    const completed = patientAppts.filter((a) => a.status === 'completed').length;
    const cancelled = patientAppts.filter((a) => a.status === 'cancelled').length;
    const noShows = patientAppts.filter((a) => a.status === 'cancelled' && !a.intake_completed).length;
    const completionRate = total > 0 ? completed / total : 0;
    const cancelRate = total > 0 ? cancelled / total : 0;
    const noShowRate = total > 0 ? noShows / total : 0;

    return [total, completed, cancelled, noShows, completionRate, cancelRate, noShowRate];
  },
};

export const paymentFeatures: FeatureExtractor = {
  name: 'payments',
  extract(ctx: RuleContext): number[] {
    const patientPayments = ctx.customerId
      ? ctx.payments.filter((p) => p.patient_id === ctx.customerId)
      : ctx.payments;

    const total = patientPayments.length;
    const completed = patientPayments.filter((p) => p.status === 'completed').length;
    const failed = patientPayments.filter((p) => p.status === 'failed').length;
    const totalRevenue = patientPayments
      .filter((p) => p.status === 'completed')
      .reduce((sum, p) => sum + p.amount, 0);
    const avgPayment = completed > 0 ? totalRevenue / completed : 0;

    return [total, completed, failed, Math.round(totalRevenue * 100) / 100, Math.round(avgPayment * 100) / 100];
  },
};

export const engagementFeatures: FeatureExtractor = {
  name: 'engagement',
  extract(ctx: RuleContext): number[] {
    const userLogs = ctx.customerId
      ? ctx.auditLogs.filter((l) => l.user_id === ctx.customerId)
      : ctx.auditLogs;

    const loginCount = userLogs.filter((l) => l.action === 'login').length;
    const lastLogin = userLogs
      .filter((l) => l.action === 'login')
      .sort((a, b) => new Date(b.created_at).getTime() - new Date(a.created_at).getTime())[0];

    const daysSinceLogin = lastLogin
      ? Math.round((Date.now() - new Date(lastLogin.created_at).getTime()) / (24 * 60 * 60 * 1000))
      : -1;

    return [loginCount, daysSinceLogin, userLogs.length];
  },
};

export const allFeatureExtractors: FeatureExtractor[] = [
  appointmentFeatures,
  paymentFeatures,
  engagementFeatures,
];
